pkg/httpserver/middleware_request_context.go

```diff
@@ -0,0 +1,23 @@
+package httpserver
+
+import (
+	"context"
+
+	"github.com/gin-gonic/gin"
+)
+
+type ContextKey string
+
+const (
+	RequestContextKey ContextKey = "http-request"
+)
+
+func ContextWithRequestMiddleware(c *gin.Context) {
+	ctx := context.WithValue(c.Request.Context(), RequestContextKey, c)
+	c.Request = c.Request.WithContext(ctx)
+	c.Next()
+}
+
+func GetRequestFromContext(ctx context.Context) *gin.Context {
```

Comment:
I believe this should be fallible if you for some reason pass in a context without the request.